src/v3/providers/redis.ts

```diff
@@ -0,0 +1,203 @@
+        cryptoMapDate[cryptoPair] = Number(mul(rate, fiatMultiplier))
+        allResults.set(normalizedDate, cryptoMapDate)
+      }
+    }
+
+    const out = expandReturnedCryptoRates(
+      requestedRates,
+      ONE_MINUTE,
+      allResults
+    )
+
+    return {
+      foundRates: out.foundRates,
+      requestedRates: out.requestedRates
+    }
+  },
+  getFiatRates: async ({ targetFiat, requestedRates }) => {
+    const rateBuckets = reduceRequestedFiatRates(
+      requestedRates,
+      TWENTY_FOUR_HOURS
+    )
+
+    const allResults: RateBuckets = new Map()
+    for (const [date, fiatPairs] of rateBuckets.entries()) {
+      let fiatMultiplier = '1'
+      if (targetFiat !== 'USD') {
+        const fiatDate = normalizeDate(date, TWENTY_FOUR_HOURS)
+        const fiatRedisKey = `rates_data:${fiatDate}:fiat`
+        const [targetFiatRate] = await hmgetAsync(fiatRedisKey, targetFiat)
+        if (targetFiatRate == null) continue
+        fiatMultiplier = targetFiatRate
+      }
+
+      const normalizedDate = normalizeDate(date, TWENTY_FOUR_HOURS)
+      const fiatRedisKey = `rates_data:${normalizedDate}:fiat`
+      const fiatPairsArray = Array.from(fiatPairs.values())
+      const rates = await hmgetAsync(fiatRedisKey, fiatPairsArray)
+
+      for (let i = 0; i < fiatPairsArray.length; i++) {
+        const rate = rates[i]
+        if (rate == null) continue
+        const fiatPair = fiatPairsArray[i]
+        const fiatMapDate = allResults.get(normalizedDate) ?? {}
+        fiatMapDate[fiatPair] = Number(mul(rate, fiatMultiplier))
+        allResults.set(normalizedDate, fiatMapDate)
+      }
+    }
+
+    const out = expandReturnedFiatRates(
+      requestedRates,
+      TWENTY_FOUR_HOURS,
+      allResults
+    )
+
+    return {
+      foundRates: out.foundRates,
+      requestedRates: out.requestedRates
+    }
+  },
+  updateRates: async (params: UpdateRatesParams): Promise<void> => {
+    if (params.targetFiat !== 'USD') {
+      return
+    }
+
+    if (params.crypto.size === 0 && params.fiat.size === 0) {
+      return
+    }
+
+    const cryptoRateBuckets = groupCryptoRatesByTime(params.crypto, ONE_MINUTE)
+    for (const [date, cryptoRates] of cryptoRateBuckets.entries()) {
+      const cryptoDate = normalizeDate(date, ONE_MINUTE)
+      const cryptoRedisKey = `rates_data:${cryptoDate}:crypto`
+      await hsetAsync(cryptoRedisKey, cryptoRates)
+    }
+
+    const fiatRateBuckets = groupFiatRatesByTime(params.fiat, TWENTY_FOUR_HOURS)
+    for (const [date, fiatRates] of fiatRateBuckets.entries()) {
+      const fiatDate = normalizeDate(date, TWENTY_FOUR_HOURS)
+      const fiatRedisKey = `rates_data:${fiatDate}:fiat`
+      await hsetAsync(fiatRedisKey, fiatRates)
+    }
+  },
+  engines: []
+}
+
+// This function breaks apart the requested rates into buckets of the given interval.
+type DateBuckets = Map<string, { [id: string]: number }>
+export const groupCryptoRatesByTime = (
```

Comment:
this is exactly `reduceRequestedCryptoRates`. Just use that and convert the Map to an Object before calling `hsetAsync`

Same for `groupFiatRatesByTime`

Reply:
They're 95% the same. reduceRequestedCryptoRates returns ids and this near duplicate function returns rates.